bigtable-hbase-parent/bigtable-hbase-integration-tests-common/src/main/java/com/google/cloud/bigtable/hbase/AbstractTestSnapshot.java

```diff
@@ -0,0 +1,229 @@
+import org.apache.hadoop.hbase.HTableDescriptor;
+import org.apache.hadoop.hbase.TableName;
+import org.apache.hadoop.hbase.client.Admin;
+import org.apache.hadoop.hbase.client.Put;
+import org.apache.hadoop.hbase.client.Result;
+import org.apache.hadoop.hbase.client.ResultScanner;
+import org.apache.hadoop.hbase.client.Scan;
+import org.apache.hadoop.hbase.client.Table;
+import org.apache.hadoop.hbase.protobuf.generated.HBaseProtos.SnapshotDescription;
+import org.apache.hadoop.hbase.util.Bytes;
+import org.junit.After;
+import org.junit.Assert;
+import org.junit.Test;
+import org.junit.experimental.categories.Category;
+
+/**
+ * <p>AbstractTestSnapshot class.</p>
+ * 
+ * @author rupeshit
+ *
+ */
+@SuppressWarnings("deprecation")
+public abstract class AbstractTestSnapshot extends AbstractTest {
+
+  protected final byte[] QUALIFIER = dataHelper.randomData("TestSnapshots");
+  
+  protected TableName tableName = sharedTestEnv.newTestTableName();
+  protected final TableName anotherTableName = sharedTestEnv.newTestTableName();
+  // The maximum size of a table id or snapshot id is 50. newTestTableName().size() can approach 50.
+  // Make sure that the snapshot name and cloned table are within the 50 character limit
+  protected final String snapshotName = tableName.getNameAsString().substring(40) + "_snp";
+  protected final String anotherSnapshotName = 
+      anotherTableName.getNameAsString().substring(40) + "_snp";
+  protected final TableName clonedTableName =
+      TableName.valueOf(tableName.getNameAsString().substring(40) + "_clone");
+  protected HTableDescriptor descriptor;
+  
+  @After
+  public void cleanup() {
+    if (sharedTestEnv.isBigtable() && !enableTestForBigtable()) {
+      return;
+    }
+    try (Admin admin = getConnection().getAdmin()) {
+      delete(admin, tableName);
+      delete(admin, anotherTableName);
+      delete(admin, clonedTableName);
+      for (SnapshotDescription snapDesc : admin.listSnapshots(snapshotName + ".*")) {
+        admin.deleteSnapshot(snapDesc.getName());
+      }
+      for (SnapshotDescription snapDesc : admin.listSnapshots(anotherSnapshotName + ".*")) {
+        admin.deleteSnapshot(snapDesc.getName());
+      }
+    } catch (Exception e) {
+      e.printStackTrace();
+    }
+  }
+
+  protected boolean enableTestForBigtable() {
+    return false;
+  }
+
+  private void delete(Admin admin, TableName tableName) throws IOException {
+    if (admin.tableExists(tableName)) {
+      admin.disableTable(tableName);
+      admin.deleteTable(tableName);
+    }
+  }
+
+  @Test
+  @Category(KnownGap.class)
+  public void testSnapshot() throws IOException {
+    if (sharedTestEnv.isBigtable() && !enableTestForBigtable()) {
+      return;
+    }
+    try (Admin admin = getConnection().getAdmin()) {
+      createTable(tableName);
+
+      Map<String, Long> values = createAndPopulateTable(tableName);
+      checkSnapshotCount(admin, 0);
+      admin.snapshot(snapshotName, tableName);
```

Comment:
please create an abstract method `snapshot(snapshotName, tableName)` that either performs a synchronous or async call.